internal/cli/server_upgrade.go

```diff
@@ -0,0 +1,363 @@
+			ctxName,
+			clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	resp, err := client.GetVersionInfo(ctx, &empty.Empty{})
+	if err != nil {
+		c.ui.Output(
+			"Error retrieving server version info: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle())
+		return 1
+	}
+
+	initServerVersion := resp.Info.Version
+
+	s.Update("Context %q validated and connected successfully.", ctxName)
+	s.Done()
+
+	s = sg.Add("Starting server snapshots")
+
+	// Snapshot server before upgrade
+	if !c.skipSnapshot {
+		s.Update("Taking server snapshot before upgrading")
+
+		snapshotName := fmt.Sprintf("%s-%d", defaultSnapshotName, time.Now().Unix())
+		if c.snapshotName != "" {
+			snapshotName = fmt.Sprintf("%s-%d", c.snapshotName, time.Now().Unix())
+		}
+
+		s.Update("Taking snapshot of server with name: '%s'", snapshotName)
+		w, closer, err := c.initWriter(snapshotName)
+		if err != nil {
+			s.Update("Failed to take server snapshot")
+			s.Status(terminal.StatusError)
+			s.Done()
+
+			c.ui.Output(fmt.Sprintf("Error opening output: %s", err), terminal.WithErrorStyle())
+			return 1
+		}
+
+		if closer != nil {
+			defer closer.Close()
+		}
+
+		if err = clisnapshot.WriteSnapshot(c.Ctx, c.project.Client(), w); err != nil {
+			s.Update("Failed to take server snapshot")
+			s.Status(terminal.StatusError)
+			s.Done()
+
+			c.ui.Output(fmt.Sprintf("Error generating Snapshot: %s", err), terminal.WithErrorStyle())
+			return 1
+		}
+
+		s.Update("Snapshot of server written to: '%s'", snapshotName)
+		s.Done()
+	} else {
+		s.Update("Server snapshot disabled on request, this means no snapshot will be taken before upgrades")
+		s.Status(terminal.StatusWarn)
+		s.Done()
+		log.Warn("Server snapshot disabled on request from user, skipping")
+	}
+
+	c.ui.Output("Waypoint server will now upgrade from version %q",
+		initServerVersion)
+
+	c.ui.Output("Upgrading...", terminal.WithHeaderStyle())
+
+	// Upgrade in place
+	result, err := p.Upgrade(ctx, &serverinstall.InstallOpts{
+		Log: log,
+		UI:  c.ui,
+	}, originalCfg.Server)
+	if err != nil {
+		c.ui.Output(
+			"Error upgrading server on %s: %s", c.platform, clierrors.Humanize(err),
+			terminal.WithErrorStyle())
+
+		c.ui.Output(upgradeFailHelp)
+
+		return 1
+	}
+
+	contextConfig := result.Context
+	advertiseAddr := result.AdvertiseAddr
+	httpAddr := result.HTTPAddr
+
+	originalCfg.Server.Address = contextConfig.Server.Address
+
+	// Connect
+	log.Info("connecting to the server so we can verify the server upgrade", "addr", originalCfg.Server.Address)
+	conn, err = serverclient.Connect(ctx,
+		serverclient.FromContextConfig(originalCfg),
+		serverclient.Timeout(5*time.Minute),
+	)
+	if err != nil {
+		c.ui.Output(
+			"Error connecting to server: %s\n\n%s",
+			clierrors.Humanize(err),
+			errInstallRunning,
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+	client = pb.NewWaypointClient(conn)
+
+	resp, err = client.GetVersionInfo(ctx, &empty.Empty{})
```

Comment:
I recommend setting the new context prior to making this call, to limit things that can fail before we persist the context.